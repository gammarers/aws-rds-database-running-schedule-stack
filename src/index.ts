import { ResourceAutoNaming, ResourceDefaultNaming, ResourceNaming, ResourceNamingType } from '@gammarers/aws-resource-naming';
import { SNSSlackMessageLambdaSubscription } from '@gammarers/aws-sns-slack-message-lambda-subscription';
import * as cdk from 'aws-cdk-lib';
import { Stack, StackProps } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as scheduler from 'aws-cdk-lib/aws-scheduler';
import * as sns from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';
import { RunningControlStateMachine } from './resources/running-control-state-machine';

export { ResourceAutoNaming, ResourceDefaultNaming, ResourceNamingType };

export interface ResourceCustomNaming {
  readonly type: ResourceNamingType.CUSTOM;
  readonly notificationTopicName: string;
  readonly notificationTopicDisplayName: string;
  readonly stateMachineName: string;
  readonly stateMachineRoleName: string;
  readonly schedulerRoleName: string;
  readonly startScheduleName: string;
  readonly stopScheduleName: string;
}

export type ResourceNamingOption = ResourceDefaultNaming | ResourceAutoNaming | ResourceCustomNaming;

export interface ScheduleProperty {
  readonly timezone: string;
  readonly minute?: string;
  readonly hour?: string;
  readonly week?: string;
}

export interface TargetResourceProperty {
  readonly tagKey: string;
  readonly tagValues: string[];
}

export interface Slack {
  readonly webhookSecretName: string;
}

export interface NotificationsProperty {
  readonly emails?: string[];
  readonly slack?: Slack;
}

export interface RDSDatabaseRunningScheduleStackProps extends StackProps {
  readonly targetResource: TargetResourceProperty;
  readonly enableScheduling?: boolean;
  readonly stopSchedule?: ScheduleProperty;
  readonly startSchedule?: ScheduleProperty;
  readonly notifications?: NotificationsProperty;
  readonly resourceNamingOption?: ResourceNamingOption;
}

export class RDSDatabaseRunningScheduleStack extends Stack {
  constructor(scope: Construct, id: string, props: RDSDatabaseRunningScheduleStackProps) {
    super(scope, id, props);

    // ðŸ‘‡ Create random 8 length string
    const random = ResourceNaming.createRandomString(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`);
    // ðŸ‘‡ Definition auto naming
    const autoNaming = {
      notificationTopicName: `rds-db-running-schedule-notification-${random}-topic`,
      notificationTopicDisplayName: 'RDS DB Running Schedule Notification Topic',
      stateMachineName: `rds-db-running-schedule-${random}-state-machine`,
      stateMachineRoleName: `rds-db-running-schedule-${random}-state-machine-role`,
      schedulerRoleName: `rds-db-running-scheduler-${random}-exec-role`,
      startScheduleName: `rds-database-running-stop-${random}-schedule`,
      stopScheduleName: `rds-database-running-start-${random}-schedule`,
    };
    // ðŸ‘‡ã€€final naming
    const names = ResourceNaming.naming(autoNaming, props.resourceNamingOption as ResourceNaming.ResourceNamingOption);


    // ðŸ‘‡ SNS Topic for notifications
    const topic: sns.Topic = new sns.Topic(this, 'NotificationTopic', {
      topicName: names.notificationTopicName,
      displayName: names.notificationTopicDisplayName,
    });

    // ðŸ‘‡ Subscribe an email endpoint to the topic
    const emails = props.notifications?.emails ?? [];
    for (const [index, value] of emails.entries()) {
      new sns.Subscription(this, `SubscriptionEmail${index.toString().padStart(3, '0')}`, {
        topic,
        protocol: sns.SubscriptionProtocol.EMAIL,
        endpoint: value,
      });
    }

    // ðŸ‘‡ Subscription slack webhook
    if (props.notifications?.slack) {
      new SNSSlackMessageLambdaSubscription(this, 'SNSSlackMessageLambdaSubscription', {
        topic,
        slackWebhookSecretName: props.notifications.slack.webhookSecretName,
      });
    }


    // ðŸ‘‡ StepFunctions State Machine
    const stateMachine = new RunningControlStateMachine(this, 'StateMachine', {
      stateMachineName: names.stateMachineName,
      notificationTopic: topic,
    });

    // ðŸ‘‡ rename role name & description.
    if (names.stateMachineRoleName) {
      const role = stateMachine.node.findChild('Role') as iam.Role;
      const cfnRole = role.node.defaultChild as iam.CfnRole;
      cfnRole.addPropertyOverride('RoleName', names.stateMachineRoleName);
      cfnRole.addPropertyOverride('Description', 'RDS DB Running machine role.');
      const policy = role.node.findChild('DefaultPolicy') as iam.Policy;
      const cfnPolicy = policy.node.defaultChild as iam.CfnPolicy;
      cfnPolicy.addPropertyOverride('PolicyName', `rds-db-running-schedule-${random}-state-machine-policy`);
    }

    // ðŸ‘‡ StateMachine Exec Role of Scheduler
    const schedulerExecRole = new iam.Role(this, 'SchedulerExecutionRole', {
      roleName: names.schedulerRoleName,
      description: 'RDS DB Running scheduler role',
      assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
      inlinePolicies: {
        'state-machine-exec-policy': new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'states:StartExecution',
              ],
              resources: [
                stateMachine.stateMachineArn,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡ Schedule state
    const sheduleState: string = (() => {
      if (props.enableScheduling === undefined || props.enableScheduling) {
        return 'ENABLED';
      } else {
        return 'DISABLED';
      }
    })();

    // ðŸ‘‡ Stop Schedule expression
    const stopScheduleExpression: string = (() => {
      // default: weekday 21:10
      const minute: string = props.stopSchedule?.minute ?? '10';
      const hour: string = props.stopSchedule?.hour ?? '21';
      const week: string = props.stopSchedule?.week ?? 'MON-FRI';
      return `cron(${minute} ${hour} ? * ${week} *)`;
    })();

    // ðŸ‘‡ Start Schedule expression
    const startScheduleExpression: string = (() => {
      // default: weekday 07:50
      const minute: string = props.startSchedule?.minute ?? '50';
      const hour: string = props.startSchedule?.hour ?? '7';
      const week: string = props.startSchedule?.week ?? 'MON-FRI';
      return `cron(${minute} ${hour} ? * ${week} *)`;
    })();

    // ðŸ‘‡ Stop DB schedule
    new DBRuningSchedule(this, 'StopDatabaseRunningSchedule', {
      name: names.startScheduleName,
      description: 'stop database(instance/cluster) running stop schedule.',
      sheduleState,
      timezone: props.stopSchedule?.timezone ?? 'UTC',
      expression: stopScheduleExpression,
      target: {
        stateMachineArn: stateMachine.stateMachineArn,
        roleArn: schedulerExecRole.roleArn,
        resourceTag: {
          key: props.targetResource.tagKey,
          values: props.targetResource.tagValues,
        },
        input: { mode: 'Stop' },
      },
    });

    // ðŸ‘‡ Start DB schedule
    new DBRuningSchedule(this, 'StartDatabaseRunningSchedule', {
      name: names.stopScheduleName,
      description: 'start db instance schedule.',
      sheduleState,
      timezone: props.startSchedule?.timezone ?? 'UTC',
      expression: startScheduleExpression,
      target: {
        stateMachineArn: stateMachine.stateMachineArn,
        roleArn: schedulerExecRole.roleArn,
        resourceTag: {
          key: props.targetResource.tagKey,
          values: props.targetResource.tagValues,
        },
        input: { mode: 'Start' },
      },
    });

  }
}

interface DBRuningScheduleProps {
  name?: string;
  description: string;
  sheduleState: string;
  timezone: string;
  expression: string;
  target: {
    stateMachineArn: string;
    roleArn: string;
    resourceTag: {
      key: string;
      values: string[];
    };
    input: {
      mode: string;
    };
  };
}

class DBRuningSchedule extends scheduler.CfnSchedule {
  constructor(scope: Construct, id: string, props: DBRuningScheduleProps) {
    super(scope, id, {
      name: props.name,
      description: props.description,
      state: props.sheduleState,
      //groupName: scheduleGroup.name, // default
      flexibleTimeWindow: {
        mode: 'OFF',
      },
      scheduleExpressionTimezone: props.timezone,
      scheduleExpression: props.expression,
      target: {
        arn: props.target.stateMachineArn,
        roleArn: props.target.roleArn,
        input: JSON.stringify({
          Params: {
            TagKey: props.target.resourceTag.key,
            TagValues: props.target.resourceTag.values,
            Mode: props.target.input.mode,
          },
        }),
        retryPolicy: {
          maximumEventAgeInSeconds: 60,
          maximumRetryAttempts: 0,
        },
      },
    });
  }
}
