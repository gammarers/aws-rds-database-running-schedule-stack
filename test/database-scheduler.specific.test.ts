import { App } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { RdsDatabaseRunningScheduleStack } from '../src';

describe('RdsDatabaseRunningScheduler Specific Testing', () => {

  describe('specify schedule', () => {
    const app = new App();

    const stack = new RdsDatabaseRunningScheduleStack(app, 'RdsDatabaseRunningScheduleStack', {
      targetResource: {
        tagKey: 'WorkHoursRunning',
        tagValues: ['YES'],
      },
      enableScheduling: true,
      startSchedule: {
        timezone: 'Asia/Tokyo',
        minute: '55',
        hour: '8',
        week: 'MON-FRI',
      },
      stopSchedule: {
        timezone: 'Asia/Tokyo',
        minute: '5',
        hour: '19',
        week: 'MON-FRI',
      },
    });

    const template = Template.fromStack(stack);

    it('Should have Schedule 4 exist', async () => {
      template.resourceCountIs('AWS::Scheduler::Schedule', 4);
    });

    it('Should have Schedule default setting', async () => {
      template.allResourcesProperties('AWS::Scheduler::Schedule', {
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        Target: {
          Arn: {
            Ref: Match.stringLikeRegexp('StateMachine'),
          },
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole'),
              'Arn',
            ],
          },
          Input: Match.anyValue(),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        },
      });
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot();
    });

  });
});