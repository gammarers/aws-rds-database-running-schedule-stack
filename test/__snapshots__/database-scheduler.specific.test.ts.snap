// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RdsDatabaseRunningScheduler Specific Testing specify schedule Should match snapshot: specify 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SchedulerExecutionRole0AF092B7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "rds:StopDBInstance",
                    "rds:StartDBInstance",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:rds:*:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":db:*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "rds-instance-stop-policy",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "rds:StopDBCluster",
                    "rds:StartDBCluster",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:rds:*:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":cluster:*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "rds-cluster-stop-policy",
          },
        ],
        "RoleName": "stop-db-schedule-f5b3cbde-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StartSchedule675DAB97": {
      "Properties": {
        "Description": "auto start db instance(db-instance-1a) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-start-db-instance-675dab97-schedule",
        "ScheduleExpression": "cron(55 8 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:rds:startDBInstance",
          "Input": "{"DbInstanceIdentifier":"db-instance-1a"}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StartSchedule940DBDA1": {
      "Properties": {
        "Description": "auto start db cluster(db-cluster-1a) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-start-db-cluster-940dbda1-schedule",
        "ScheduleExpression": "cron(55 8 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:rds:startDBCluster",
          "Input": "{"DbClusterIdentifier":"db-cluster-1a"}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StopSchedule675DAB97": {
      "Properties": {
        "Description": "auto stop db instance(db-instance-1a) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-stop-db-instance-675dab97-schedule",
        "ScheduleExpression": "cron(5 19 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:rds:stopDBInstance",
          "Input": "{"DbInstanceIdentifier":"db-instance-1a"}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StopSchedule940DBDA1": {
      "Properties": {
        "Description": "auto stop db cluster(db-cluster-1a) schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "auto-stop-db-cluster-940dbda1-schedule",
        "ScheduleExpression": "cron(5 19 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": {
          "Arn": "arn:aws:scheduler:::aws-sdk:rds:stopDBCluster",
          "Input": "{"DbClusterIdentifier":"db-cluster-1a"}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
